### PR TITLE
[HW][Comb] Canonicalize muxes of arrays

### PR DESCRIPTION
**Pull request by nandor**
Array create operations which build arrays from muxes with identical conditions are turned into muxes of arrays.
An inverse transformation scalarizes the muxes if their array arguments are uniform.

### CODE REVIEW

File: lib/Dialect/Comb/CombFolds.cpp [hunk 1]
```diff
@@ -2144,7 +2144,38 @@ static bool foldCommonMuxOperation(MuxOp mux, Operation *trueOp,
...
+  auto trueVec = op.getTrueValue().getDefiningOp<hw::ArrayCreateOp>();
+  auto falseVec = op.getFalseValue().getDefiningOp<hw::ArrayCreateOp>();
+  if (!trueVec || !falseVec)
+    return false;
+  if (!trueVec.isUniform() || !falseVec.isUniform())
+    return false;
+
+  auto mux = rewriter.create<MuxOp>(op.getLoc(), op.getCond(),
+                                    trueVec.getUniformElement(),
+                                    falseVec.getUniformElement());
```
Comment (uenoku) [inline lib/Dialect/Comb/CombFolds.cpp:2159]: Can we propagate two state flag here?

File: lib/Dialect/Comb/CombFolds.cpp [hunk 2]
```diff
@@ -2144,7 +2144,38 @@ static bool foldCommonMuxOperation(MuxOp mux, Operation *trueOp,
...
+
+  SmallVector<Value> values;
+  for (unsigned i = 0, n = trueVec.getInputs().size(); i < n; ++i)
+    values.push_back(mux);
```
Comment (uenoku) [inline lib/Dialect/Comb/CombFolds.cpp:2163]: Does `SmallVector<Value> values(trueVec.getInputs().size(), mux);` work?

File: lib/Dialect/Comb/CombFolds.cpp [hunk 3]
```diff
@@ -2328,9 +2359,76 @@ LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
...
+    if (!input || first.getCond() != input.getCond())
+      return false;
+  }
+
+  // Collect the true and the false branches into arrays.
+  SmallVector<Value> trues{first.getTrueValue()};
+  SmallVector<Value> falses{first.getFalseValue()};
+  SmallVector<Location> locs{first->getLoc()};
+  for (size_t i = 1, n = inputs.size(); i < n; ++i) {
+    auto input = inputs[i].getDefiningOp<comb::MuxOp>();
```
Comment (uenoku) [inline lib/Dialect/Comb/CombFolds.cpp:2395]: Can we propagate two state flag if all operands have a binary flag?

File: lib/Dialect/Comb/CombFolds.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Dialect/Comb/CombFolds.cpp:2163]: Considered that there is `comb.replicate` we might want to introduce `hw.array_repliacte` or something.

**Review by uenoku — APPROVED** [review]
The canonicalization looks good to me.

File: lib/Dialect/Comb/CombFolds.cpp (hunk 2 above)
Comment (nandor) [reply lib/Dialect/Comb/CombFolds.cpp:2163]: That would be useful, although there is some overlap with `array_create`. I think for this particular case we could consider specialising array_create of uniform arrays to store the items more efficiently.

File: lib/Dialect/Comb/CombFolds.cpp [hunk 4]
```diff
@@ -2328,9 +2356,79 @@ LogicalResult MuxOp::canonicalize(MuxOp op, PatternRewriter &rewriter) {
...
+    if (!input || first.getCond() != input.getCond())
+      return false;
+  }
+
+  // Collect the true and the false branches into arrays.
+  SmallVector<Value> trues{first.getTrueValue()};
+  SmallVector<Value> falses{first.getFalseValue()};
+  SmallVector<Location> locs{first->getLoc()};
+  bool isTwoState = true;
```
Comment (uenoku) [inline lib/Dialect/Comb/CombFolds.cpp:2391]: nit: `isTwoState = first.getTwoState()`?

**Pull request merged**